Move AI autofix logic into transformers so suggestions can be cached

lsegal opened the pull request:
My original attempt at this implementation was manually caching the autofix() step, but it seemed very messy (that impl is not provided in this PR).

This approach feels a lot cleaner because it utilizes IssueTransformer to mutate the issue. We can also hook directly into progress to provide a much more understandable inline UX.

brynary reviewed (approved):
A transformer for this concern seems fine, and I would be happy to land this.

One thing that's interesting is concurrency... I don't recall if we currently are using concurrency in our API calls for fixes, but I suspect that we will want to do that to reduce end user latency.

There's also the possibility of batching especially if we are generating fixes for multiple issues in the same file. In that case we might be able to only upload the file once instead of N times. I believe that there is a stub of this concept in the API definition which hasn't been implemented yet.

I think where we want to end up eventually is using some form of bandwidth and concurrency. So for example one API call to get all the fixes in a single file (up to a maximum), with the CLI running multiple requests to the API concurrently. (And then the API server making concurrent requests to the LLM for batch fix generation.)

These concepts are not necessarily incompatible with the transformer API, but I'd have to think what that would look like.

lsegal commented:
> One thing that's interesting is concurrency... I don't recall if we currently are using concurrency in our API calls for fixes, but I suspect that we will want to do that to reduce end user latency.

We do (did) use concurrency, and in fairness, we did have a much better handle on the max concurrency in the implementation that this PR replaces. 

The prior implementation controlled concurrency across _all_ issues, and had a max threads for generating AI fixes (it was set to 12). This means we had at most 12 concurrent calls.

The current implementation relies on existing threading in analysis, of which we parallelize invocations, but not transforms on a single invocation, so: technically there is concurrency, but we don't have much control over how much concurrency we would get out of this. If we run 100 invocations at the same time, the AI code would likely run 100 reqs concurrently. That's probably not ideal (although this is far beyond the edge case of what we would expect).

I think we (a) want to implement a batching endpoint, but also (b) probably want some concurrency limit in the API client request itself, because trying to do it at the transformer layer will be difficult. That said, we also likely want to expand the ability to run the transforms in parallel, as this directly affects the type of concurrency we can achieve, though I guess if we go for batching, this becomes less relevant (and almost unwanted).

tl;dr there's a bit to consolidate on concurrency if we merge this, but I think the behavior will be _roughly_ the same as it is today.

brynary commented:
Thanks for the additional info.

The number of concurrency invocations defaults to the number of CPUs found. It can be overridden (I believe up to a maximum), but that is rarely used. So in practice (local and on Cloud), invocations usually run with a concurrency of 2 - perhaps 32 cpus.

At the low end, which I believe is the Cloud default, fix generation will be slower that it needs to be. And at the high end it may be more concurrency than we want a single CLI to hit the server with. 

Implementing concurrency control at the API layer can prevent it from going too high, but can't get us more concurrency than the invocations concurrency.

So to generate optimal API concurrency from a transformer strategy, it seems like we would need to both run transformers concurrently _across invocations_ and also implement a concurrency maximum at the API client level (effectively a connection pool type algorithm).

Does that sound accurate? If so, do you think we could  adjust the CLI to run the transformers fully concurrently without too much effort and/or code complexity?

lsegal commented:
@brynary I just added extra concurrency to executor processing to allow for the same behavior we had before wrt. concurrency:

https://github.com/qltysh/qlty/pull/1519/commits/3b3e8fa10aef37caf53bdbfe76e3266112111d28

This expands concurrency in file_result processing, something we missed a while ago in perf optimizations and should make overall processing slightly faster-- but will have the biggest effect with the AI fixes and computationally heavy transforms. 

As a result, I added an ad-hoc concurrency limit in the AI API request network code. Right now it's localized to the Fixer struct but we could probably refactor this out to the entire client at some point in the future.

Curious your thoughts.

brynary commented:
Thanks! Sorry I missed your last comment and commit. Will review this weekend with some coffee.

brynary reviewed (commented):
OK, had a chance to take a look at this. I left a few comments inline, including some questions about some potentially lost behavior. 

RE: the concurrency controls in fixer.rs.  I am a little bit worried that they are a bit more complex than rayon pool strategy which is easier to follow. (And maybe less risk of a deadlock bug?)

Do we need all three of `attempts_per_file`, `api_concurrency_lock` and `api_concurrency_guard`, or is there potentially a way to simplify this?

Personally, I don't have a lot of faith in my own ability to manually deal with concurrency. I top out at about around the `Arc<AtomicUsize>` level, and don't really trust myself to use things like Mutex and RwLock properly without creating issues. So I try to lean on rayon.

lsegal commented on qlty-check/src/llm/fixer.rs:
```diff
@@ -1,196 +1,100 @@
-                    >= *issue.line_range().unwrap().start()
-            {
-                debug!(
-                    "Skipping issue too close to previous at {}:{}",
```
This is not accidental. Because this is only generating suggestions and not applying them, there's no value to choosing on locality. When prompted, a user might choose to fix the "second" suggestion that would have been skipped by this check but not the first.

lsegal commented on qlty-check/src/llm/fixer.rs:
```diff
@@ -1,196 +1,100 @@
-    results: Results,
     r#unsafe: bool,
-    progress: bool,
-    pre_transformers: Vec<Box<dyn IssueTransformer>>,
```
Moving Fixer into the regular transformers stack means we're already running CheckFilters:

https://github.com/qltysh/qlty/blob/db7fd1e53911c5432efcd84505838577a8f71a06/qlty-check/src/executor.rs#L132-L134

lsegal commented on qlty-check/src/executor.rs:
```diff
@@ -635,15 +640,16 @@ fn run_invocation(
                     MAX_ISSUES_PER_FILE
                 );
-                issue_limit_reached.insert(PathBuf::from(&file_result.path));
+                issue_limit_reached.lock().unwrap().insert(PathBuf::from(&file_result.path));
```
This is a little more complicated since PoisonErrors cannot be returned in this fashion. They generally have to be boxed / replaced, thus losing the original error object.

lsegal replied:
Also: this is inside of a closure that does not return a Result, and thus we cannot use ? in this specific spot.

lsegal commented:
> Do we need all three of attempts_per_file, api_concurrency_lock and api_concurrency_guard, or is there potentially a way to simplify this?

attempts_per_file is just a synchronized shared structure, it would need to be inside of a mutex regardless of whether we used rayon or not (since it needs to be shared across threads and thus Sync). This specific change is actually _because_ we started using rayon instead of a single threaded loop (since the entire Fixer struct now runs in the context of threaded transformers).

The api_concurrency stuff is separate (below)

> RE: the concurrency controls in fixer.rs. I am a little bit worried that they are a bit more complex than rayon pool strategy which is easier to follow. (And maybe less risk of a deadlock bug?)

This is probably a good idea in general.
